gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionTransformer.scala

```diff
@@ -16,17 +16,74 @@
  */
 package org.apache.gluten.expression
 
-import org.apache.gluten.substrait.expression.ExpressionNode
+import org.apache.gluten.substrait.expression.{ExpressionBuilder, ExpressionNode}
 
-import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.expressions.{Expression, Literal}
 import org.apache.spark.sql.types.DataType
 
-trait ExpressionTransformer {
-  def doTransform(args: java.lang.Object): ExpressionNode
-  def dataType: DataType
-}
+import scala.collection.JavaConverters._
+
+// ==== Expression transformer basic interface start ====
 
-trait ExpressionTransformerWithOrigin extends ExpressionTransformer {
+trait ExpressionTransformer {
+  def substraitExprName: String
+  def children: Seq[ExpressionTransformer]
   def original: Expression
   def dataType: DataType = original.dataType
+  def nullable: Boolean = original.nullable
+
+  def doTransform(args: java.lang.Object): ExpressionNode = {
+    val functionMap = args.asInstanceOf[java.util.HashMap[String, java.lang.Long]]
+    // TODO: the funcName seems can be simplified to `substraitExprName`
+    val funcName: String =
+      ConverterUtils.makeFuncName(substraitExprName, original.children.map(_.dataType))
+    val functionId = ExpressionBuilder.newScalarFunction(functionMap, funcName)
+    val childNodes = children.map(_.doTransform(args)).asJava
+    val typeNode = ConverterUtils.getTypeNode(dataType, nullable)
+    ExpressionBuilder.makeScalarFunction(functionId, childNodes, typeNode)
+  }
+}
+
+trait LeafExpressionTransformer extends ExpressionTransformer {
+  final override def children: Seq[ExpressionTransformer] = Nil
+}
+
+trait UnaryExpressionTransformer extends ExpressionTransformer {
+  def child: ExpressionTransformer
+  final override def children: Seq[ExpressionTransformer] = child :: Nil
+}
+
+trait BinaryExpressionTransformer extends ExpressionTransformer {
+  def left: ExpressionTransformer
+  def right: ExpressionTransformer
+  final override def children: Seq[ExpressionTransformer] = left :: right :: Nil
+}
+
+// ==== Expression transformer basic interface end ====
+
+case class GenericExpressionTransformer(
```

Comment:
Most expression can leverage this transformer rather than make a custom expression. Take an example `MakeDecimal`, it is a unary expression in vanilla Spark but native requires one more arg `nullOnOverflow`, then we can pass one more child `LiteralTransformer` to 
`GenericExpressionTransformer`:
```scala
GenericExpressionTransformer(
  substraitExprName,
  Seq(
    replaceWithExpressionTransformerInternal(m.child, attributeSeq, expressionsMap),
    LiteralTransformer(m.nullOnOverflow)),
  m
)
```